kiali/kiali-ui
# Workload detail actions + 3scale extension changes

Fixes https://github.com/kiali/kiali/issues/2860
Fixes https://github.com/kiali/kiali/issues/2870

Requires https://github.com/kiali/kiali/pull/2936

In summary:
- Remove 3scale extension a from main navigation menu as 3scale is going to be configured using Istio Mixer resources: Handler, Instance, Rules.
- Create an specific action menu for 3scale config under Istio Config page.
- Introduce a 3scale config page that maps 3scale concepts into Istio Mixer config (3scale account and authorization).
- Move 3scale link action from Service to Workload details.
- Remove 3scale specific models.

How to test:

- When 3scale extension is disabled:
```
extensions:
  threescale:
    adapter_name: threescale
    adapter_port: 3333
    adapter_service: threescale-istio-adapter
    enabled: false
    template_name: threescale-authorization
  iter_8:
    enabled: false
```

There should not be a 3scale action in the Istio Config List -> Actions 
![image](https://user-images.githubusercontent.com/1662329/86486611-04007080-bd5c-11ea-92f7-91dd1d73d1e3.png)

And Workload details shouldn't show any Action:
![image](https://user-images.githubusercontent.com/1662329/86486670-2f835b00-bd5c-11ea-9d2d-ce5e5c650e06.png)

(Note, while writting this I see that this PR is not complete, it should disable/hide workload actions if there is none)

- When 3scale extension is enabled:
![image](https://user-images.githubusercontent.com/1662329/86486809-943eb580-bd5c-11ea-9e52-31970b354fb4.png)

3scale navigation menu doesn't exist anymore, now it's implemented with a new Istio Contig List -> Actions -> Create 3scale Config form

A Create 3scale Config form allows to create 3scale Accounts mapped with a Istio Handler resource:
![image](https://user-images.githubusercontent.com/1662329/86487004-38c0f780-bd5d-11ea-9824-c357242ae5d7.png)

And 3scale Authorization mapped with a pair of Istio Instance+Rule:
![image](https://user-images.githubusercontent.com/1662329/86487031-51311200-bd5d-11ea-8441-1fa9d775b4b6.png)

Business rationale about this is detailed in the Issue https://github.com/kiali/kiali/issues/2860

At the end this form creates standar Istio Mixer objects wtith the 3scale configuration mapped into them:
![image](https://user-images.githubusercontent.com/1662329/86487371-298e7980-bd5e-11ea-85d5-c71b05ba59d3.png)

A "3scale credential" is a convention name to refer a pair of Istio Instance+Rule previously created in the form.

Now, under the Workload Details there is a new action to link a workload with 3scale:
![image](https://user-images.githubusercontent.com/1662329/86487456-6fe3d880-bd5e-11ea-82d0-8c8719c9094f.png)

The user needs to introduce a "3scale service id" identifier which is previously generated in 3scale system and user needs to know:
![image](https://user-images.githubusercontent.com/1662329/86487508-96097880-bd5e-11ea-905b-391772df1ecb.png)

When workload is "linked" with 3scale, additional labels will be added to the workload and propagated into Pods:
![image](https://user-images.githubusercontent.com/1662329/86487578-be917280-bd5e-11ea-83e9-29191d1eaac5.png)

(Note that a re-start of Pods is expected)

When a workload is detected to be "linked" with 3scale (checking the 3scale labels), then the "unlink" operation appears:
![image](https://user-images.githubusercontent.com/1662329/86487641-f00a3e00-bd5e-11ea-8b3c-4a737bcb8299.png)

Note that when modifying the labels and during the Pods restart, Kiali can fetch that situation:
![image](https://user-images.githubusercontent.com/1662329/86487722-28118100-bd5f-11ea-838c-7fb443da6580.png)

So, this PRs refactor partially how Workload is refreshed.

One known defect/pending issue to address is to refactor the Logs tab to allow to refresh the workload at tab level and move the controls of refresh in the same location that others tab:

![image](https://user-images.githubusercontent.com/1662329/86487800-6444e180-bd5f-11ea-9f1b-bc8e755be41f.png)

See, that the logs tab has the old pod container instead the new one, the workaround is to refresh the page.
As this PR is growing, perhaps that change can be done in a following PR, but it's a topic to comment.
